src/main/java/org/opensearch/agent/tools/DataDistributionTool.java

```diff
@@ -0,0 +1,1304 @@
+            Object value = getFlattenedValue(doc, field);
+            if (value != null) {
+                String strValue = String.valueOf(value);
+                counts.merge(strValue, 1, Integer::sum);
+            }
+        }
+
+        return counts.entrySet().stream().collect(Collectors.toMap(Map.Entry::getKey, entry -> (double) entry.getValue() / data.size()));
+    }
+
+    /**
+     * Calculates maximum difference between selection and baseline distributions
+     *
+     * @param selectionDist Selection period distribution
+     * @param baselineDist Baseline period distribution
+     * @return Maximum difference value across all field values
+     */
+    private double calculateMaxDifference(Map<String, Double> selectionDist, Map<String, Double> baselineDist) {
+        Set<String> allKeys = new HashSet<>(selectionDist.keySet());
+        allKeys.addAll(baselineDist.keySet());
+
+        if (allKeys.isEmpty()) {
+            return Double.NEGATIVE_INFINITY;
+        }
+        return allKeys.stream().mapToDouble(key -> {
+            double selectionVal = selectionDist.getOrDefault(key, 0.0);
+            double baselineVal = baselineDist.getOrDefault(key, 0.0);
+            return selectionVal - baselineVal;
+        }).max().orElse(Double.NEGATIVE_INFINITY);
+    }
+
+    /**
+     * Extracts field names from sample data when mapping is not available
+     *
+     * @param data Sample data to analyze
+     * @return List of field names suitable for analysis
+     */
+    private List<String> getFieldsFromData(List<Map<String, Object>> data) {
+        if (data.isEmpty()) {
+            return List.of();
+        }
+
+        Set<String> allFields = new HashSet<>();
+        for (Map<String, Object> doc : data) {
+            allFields.addAll(doc.keySet());
+        }
+
+        // Filter out timestamp and other non-useful fields
+        return allFields
+            .stream()
+            .filter(field -> !field.equals("@timestamp") && !field.equals("_id") && !field.equals("_index"))
+            .filter(field -> {
+                // Check cardinality - exclude high cardinality fields
+                Set<String> values = new HashSet<>();
+                for (Map<String, Object> doc : data) {
+                    Object value = doc.get(field);
+                    if (value != null) {
+                        values.add(String.valueOf(value));
+                    }
+                }
+                int cardinality = values.size();
+                return cardinality > 0 && cardinality <= Math.max(10, data.size() / 2);
+            })
+            .collect(Collectors.toList());
+    }
+
+    /**
+     * Gets number fields from index mapping
+     *
+     * @param index Index name
+     * @return Set of number field names
+     */
+    private Set<String> getNumberFields(String index) {
+        try {
+            GetMappingsRequest getMappingsRequest = new GetMappingsRequest().indices(index);
```

Comment:
Looks we get mappings twice in `getNumberFields` and `getUsefulFields`. Can we reuse information?